src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -262,8 +297,92 @@ public async ValueTask UnsubscribeAsync(int sid, int? maxMsgs, CancellationToken
     // only used for internal testing
     internal bool TestStallFlush() => _channelLock.Writer.TryWrite(1);
 
-    [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void ThrowOnDisconnected() => throw new NatsException("Connection hasn't been established yet.");
+    private static async Task ReaderLoopAsync(ILogger<CommandWriter> logger, ISocketConnection connection, PipeReader pipeReader, Channel<int> channelSize, CancellationToken cancellationToken)
+    {
+        try
+        {
+            while (true)
+            {
+                var result = await pipeReader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+                if (result.IsCanceled)
+                {
+                    break;
+                }
+
+                var buffer = result.Buffer;
+                var completed = buffer.Start;
+                try
+                {
+                    if (!buffer.IsEmpty)
+                    {
+                        var bufferLength = (int)buffer.Length;
+
+                        var bytes = ArrayPool<byte>.Shared.Rent(bufferLength);
+                        buffer.CopyTo(bytes);
+                        var memory = bytes.AsMemory(0, (int)buffer.Length);
+
+                        try
+                        {
+                            completed = buffer.Start;
+                            var totalSent = 0;
+                            var totalSize = 0;
+                            while (totalSent < bufferLength)
+                            {
+                                var sent = await connection.SendAsync(memory).ConfigureAwait(false);
+
+                                totalSent += sent;
+                                memory = memory[sent..];
+
+                                while (totalSize < totalSent)
+                                {
+                                    int size;
+                                    while (!channelSize.Reader.TryRead(out size))
+                                    {
+                                        await channelSize.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false);
+                                    }
+
+                                    totalSize += size;
+                                }
```

Comment:
Don't want to pop off channel before the command has been completely sent.  So I think this needs to be more like:

```
while (true) {

peek channel

if (totalSize + peek > totalSent) {
  break;
}

read channel
totalSize += size

}
```